ironfish/src/mining/manager.ts

```diff
@@ -167,26 +151,25 @@ export class MiningManager {
       return
     }
 
-    // Kick off job to creating the next empty miners fee
-    this.minersFeeCache.startCreatingEmptyMinersFee(currentBlock.header.sequence + 2, account)
+    const emptyTemplate = await this.createNewBlockTemplate(currentBlock, account, false)
+    this.metrics.mining_newEmptyBlockTemplate.add(BenchUtils.end(connectedAt))
+    this.streamBlockTemplate(currentBlock, emptyTemplate)
 
-    void this.createNewEmptyBlockTemplate(currentBlock, account)
-      .then((template: SerializedBlockTemplate) => {
-        this.metrics.mining_newEmptyBlockTemplate.add(BenchUtils.end(connectedAt))
-        this.streamBlockTemplate(currentBlock, template, 0)
-      })
-      .catch((error) => {
-        this.node.logger.info(`Discarding block template: ${ErrorUtils.renderError(error)}`)
-      })
+    // The head of the chain has changed, abort working on this template
+    if (!this.chain.head.hash.equals(currentBlock.header.hash)) {
+      return
+    }
 
-    void this.createNewBlockTemplate(currentBlock, account)
-      .then((template: SerializedBlockTemplate) => {
-        this.metrics.mining_newBlockTemplate.add(BenchUtils.end(connectedAt))
-        this.streamBlockTemplate(currentBlock, template, 1)
-      })
-      .catch((error) => {
-        this.node.logger.info(`Discarding block template: ${ErrorUtils.renderError(error)}`)
-      })
+    // Only try creating a block with transactions if there are transactions in
+    // the mempool
+    if (this.memPool.count()) {
+      const template = await this.createNewBlockTemplate(currentBlock, account)
+      this.metrics.mining_newBlockTemplate.add(BenchUtils.end(connectedAt))
+      this.streamBlockTemplate(currentBlock, template)
+    }
+
+    // Kick off job to create the next empty miners fee
+    this.minersFeeCache.startCreatingEmptyMinersFee(currentBlock.header.sequence + 2, account)
```

Comment:
would we still want to still kick off this job at the beginning? Or at least before we finish creating the full block?

Reply:
My assumption is that the order of priorities should be:
1. Get an empty block out ASAP - this is for the miners' benefit
2. Get a block with transactions out ASAP - this is for the network's benefit
3. Pre-empt the next miners fee. This is closely related to 1, since without this, 1. is delayed on the next block. If we assume the next block happens within 2-5s or whatever, then the next empty block suffers.

Thinking through this, since _most_ of the work for miners fee happens in a worker job, I think it is a good idea to move this above the full block kick-off, since they can happen in tandem. Although I do think getting the empty block sent out should be priority 1, and kicking off the job isnt quite free (although it's pretty close).

I'll move this up, good call.

Reply:
Yea I agree
1. Empty block
2. Kick off next miners fee
3. Full block